jetty-documentation/src/main/asciidoc/development/maven/jetty-maven-helloworld.adoc

```diff
@@ -32,6 +32,13 @@ Using Maven for Jetty implementations is a popular choice, but users encouraged
 Other popular tools include Ant and Gradle. 
 ____
 
+____
+[IMPORTANT]
+Various Jetty sub-projects may define configuration files with the same name.
```

Comment:
These are not jetty configuration files.  They are standard jar file features in META-INF. 

The note should say that jetty uses the features of Service loader and increasing the feature of multi-version jar. Both have files in META-INF that need to be correctly merged/renamed if creating uber or shaded jars.      
Having a link to the ServicesResourceTransformer is good.

Reply:
@gregw  I confess that you stumped me here: "jetty uses the features of Service loader and increasing the feature of multi-version jar."  It sounds like something that deserves a JSR, but I don't see one.  The best I could find is this description of the Jar file format mostly copied from an old Sun site to a newer Oracle page:
https://docs.oracle.com/javase/7/docs/technotes/guides/jar/jar.html#Provider_Configuration_File

That explains the META-INF/services directory where service provider configuration files live, but says nothing about how competing provider-configuration-files should be handled.  It doesn't specify whether they should be merged, overwritten, or if conflicts are errors.

Do you have a link to the standard that specifies provider configuration file merging?  Even the [ServicesResourceTransformer](https://maven.apache.org/plugins/maven-shade-plugin/examples/resource-transformers.html#ServicesResourceTransformer) makes it sound like a choice "to merge multiple implementations of the same interface into one service entry."

I'd really like to educate myself about this a little more before I try again to write a sentence about it.  If merging provider configuration files is a Jetty assumption or even just a common assumption among the Jar File Cognoscenti, we should be clear about that.

Gradle:
Here's an example of calling a ServicesResourceTransformer in Gradle:
https://github.com/johnrengelman/shadow/issues/105

Here's the only other scrap I could find:
https://stackoverflow.com/questions/32887966/shadow-plugin-gradle-what-does-mergeservicefiles-do

I could test out the Gradle recipe with my project and report back.  It seems that's going to be a requirement in order to build a fat jar with Gradle.

Sorry to take so long.  My non-work coding is very limited at the moment.

Reply:
If you are making a uber-jar (the more common name to "far jar"), then you merge all `META-INF/services/<name>` files across all jars together, removing duplicate lines (this last part is not terribly important for a library that uses `META-INF/services` properly.)
Each line is always a fully qualified Class filename.

Multi-Release JAR files have a JEP (the successor to JSR), it's https://openjdk.java.net/jeps/238

Reply:
I also can't find anywhere that specifies merging META-INF/services, but then there is nothing in the jar spec that talks about merging jars in general.
I think it is just emergent behaviour.  At least with META-INF/services the spec is such that conflicts are easy to resolve (by appending).  Merging the manifest is actually a much harder task as there is plenty more scope for unsolvable conflicts there.